Python/fileutils.c

```diff
@@ -1847,3 +1847,50 @@ _Py_GetLocaleconvNumeric(PyObject **decimal_point, PyObject **thousands_sep,
     PyMem_Free(oldloc);
     return res;
 }
+
+
+#ifdef MS_WINDOWS
+
+/* Determine whether a reparse point is a link type. */
+int
+_Py_is_reparse_link(const wchar_t *path, ULONG reparse_tag)
+{
+    BOOL ret;
+    BOOL result = FALSE;
+    wchar_t *mountpath = NULL;
+    size_t buflen;
+    if (reparse_tag == IO_REPARSE_TAG_SYMLINK)
+        return TRUE;
+    else if (reparse_tag == IO_REPARSE_TAG_MOUNT_POINT)
+    {
+        /* Junction point. Either a link (e.g. mklink /j) or a volume mount
+           points (e.g. mountvol.exe). Compare GetVolumePathNameW to path
+           to determine which case. If they are equal, it is a mount.
+        */
+
+        /* Volume path should be shorter than entire path */
```

Comment:
`path` could be relative or use forward slashes. You have to normalize it via `GetFullPathName`. 

Alternatively, you could read the junction target via `FSCTL_GET_REPARSE_POINT` and check for a "Volume{GUID}" name using a variation of the `MOUNTMGR_IS_VOLUME_NAME` macro (from mountmgr.h in the WDK).

Reply:
> `path` could be relative [...]

I see that `GetFullPathName` requires the GIL (relies on the process-global current directory). However, one or more of the call paths for `_Py_is_reparse_link` is wrapped in a `Py_BEGIN/END_ALLOW_THREADS`, e.g. `posix_do_stat` -> `win32_stat` -> `win32_xstat`. What is the recommended way to resolve this? Is this already an error, as `win32_xstat` calls `CreateFileW` with `path` (or does that use another method of resolving relative paths?)?

Reply:
@eryksun might correct me, but AFAIK all Win32 APIs canonicalize paths with `RtlGetFullPathName_U` internally, and `GetFullPathName` is a wrapper around that function. I don't think you should worry about GIL here.

Reply:
Don't worry about the GIL for Windows Process Environment Block (PEB) data. Whatever is critical there is already protected by locks (e.g. the loader lock). The GIL is to protect interpreter and object state. That said, there's certainly no reason to *release* the GIL to call `GetFullPathNameW`. It's a runtime library function that operates on strings using a set of rules for DOS path processing. 

@izbyshev:

As to how WinAPI functions normalize a path, `GetFullPathNameW` in particular calls `RtlGetFullPathName_U[Ex]`, which in turn calls the more commonly-used function, `RtlGetFullPathName_Ustr`. The latter takes a counted `UNICODE_STRING` and also implements some optional behavior that's not used by `GetFullPathNameW`. 

Internally a path is normalized and translated to a native NT path. The native path gets passed to an I/O system call (commonly `NtCreateFile`, `NtOpenFile`, `NtCreateNamedPipeFile`, `NtCreateMailslotFile`, `NtQueryAttributesFile`, `NtQueryFullAttributesFile`, or `NtDeleteFile`) as part of an `OBJECT_ATTRIBUTES` record. (This structure is used for all named NT objects, not just Device objects.) DOS-to-NT path translation usually uses either `RtlDosPathNameToNtPathName_U[_WithStatus]` or `RtlDosPathNameToRelativeNtPathName_U[_WithStatus]`. These functions make use of the optional features of `RtlGetFullPathName_Ustr`. This includes checking for a valid path (e.g. the DOS device "NUL" in the path "C:\Temp\NUL" is only valid if "C:\Temp" exists) and returning the DOS path type (e.g. local-device absolute "\\\\.\\" and "\\\\?\\"; UNC absolute; drive absolute, rooted, drive relative, relative). 

If a relative path is beneath the working directory in the tree, then Windows can use a handle for the working directory as the `RootDirectory` field of the `OBJECT_ATTRIBUTES` record. Otherwise it has to resolve the path to an absolute NT path that begins with the virtual DOS device directory "\\??\\" (a combined view of the caller's local DOS-device links for the logon session in "\Sessions\0\DosDevices\\[Logon ID]" and the global DOS-device links in "\GLOBAL??"). Note that this is not WinAPI "\\\\?\\", but in practice  "\\\\?\\" bypasses all other path processing, so it simply gets replaced by "\\??\\" when translated to an NT path.